### PR TITLE
[Feature][API] New restful API for workflow and schedule 

### PR DESCRIPTION
**Pull request by zhongjiajie**
CURD for workflow and schedule, different with exists
API, this new restful api only operate single resource
in each request, and return the latest. For example,
previous workflow should also need to post tasks definition
and tasks relation definition, but this patch will allow
you to create workflow without task relate information

### CODE REVIEW

**Comment by zhongjiajie** [comment]
This PR is not equal to refactored exists API as #10257 do, we also separate existing`process definition` and `schedule` into single resources CRUD(previous process definition also will submit tasks and task relation). And we make it more Restful

**Comment by zhongjiajie** [comment]
fix conflict from adding log in #11782

**Comment by zhongjiajie** [comment]
PTAL if you have time @caishunfeng @SbloodyS

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/ScheduleV2Controller.java [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+     * @return result Result
+     */
+    @ApiOperation(value = "get", notes = "GET_SCHEDULE_BY_ID_NOTES")
+    @GetMapping(value = "/{id}")
+    @ResponseStatus(HttpStatus.OK)
+    @ApiException(QUERY_SCHEDULE_LIST_ERROR)
+    @AccessLogAnnotation(ignoreRequestArgs = "loginUser")
+    public ResourceResponse getSchedules(@ApiIgnore @RequestAttribute(value = Constants.SESSION_USER) User loginUser,
+                                         @PathVariable("id") Integer id) {
+        Schedule schedule = schedulerService.getSchedules(loginUser, id);
```
Comment (caishunfeng) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/ScheduleV2Controller.java]:
```suggestion
        Schedule schedule = schedulerService.getSchedule(loginUser, id);
```

Comment (caishunfeng) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/ScheduleV2Controller.java]: Or change into multi method.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/dto/schedule/ScheduleCreateRequest.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    @ApiModelProperty(allowableValues = "HIGHEST / HIGH / MEDIUM / LOW / LOWEST", example = "MEDIUM", notes = "default MEDIUM if value not provide.")
+    private String processInstancePriority;
+
+    @ApiModelProperty(example = "worker-group-name")
+    private String workerGroup;
+
+    @ApiModelProperty(example = "environment-code")
+    private long environmentCode;
+
+    public String getScheduleParam() {
```
Comment (caishunfeng) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/dto/schedule/ScheduleCreateRequest.java:81]: Why not use json util?

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/dto/schedule/ScheduleUpdateRequest.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+                format(scheduleUpdate.getEndTime(), YYYY_MM_DD_HH_MM_SS, schedule.getTimezoneId()) +
+                "\",\"crontab\":\"" +
+                scheduleUpdate.getCrontab() +
+                "\",\"timezoneId\":\"" +
+                scheduleUpdate.getTimezoneId() +
+                "\"}";
+    }
+
+    public Schedule mergeIntoSchedule(Schedule schedule) {
+        Schedule scheduleDeepCopy = JSONUtils.parseObject(JSONUtils.toJsonString(schedule), Schedule.class);
```
Comment (caishunfeng) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/dto/schedule/ScheduleUpdateRequest.java]: use BeanUtils.cloneBean

**Review by songjianet — DISMISSED** [review]
LGTM

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java [hunk 4]
```diff
@@ -67,6 +70,15 @@ Map<String, Object> createProcessDefinition(User loginUser,
...
                                                 ProcessExecutionTypeEnum executionType);
 
+    /**
+     * create process definition V2
+     *
+     * @param loginUser login user
+     * @param workflowCreateRequest the new workflow object will be created
+     * @return New ProcessDefinition object created just now
+     */
+    ProcessDefinition createProcessDefinitionV2(User loginUser, WorkflowCreateRequest workflowCreateRequest);
```
Comment (caishunfeng) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java]: What's the different between v1 and v2? Maybe v2 it's not a good guide for developers if v1 can not be replaced with v2.

**Review by caishunfeng — CHANGES_REQUESTED** [review]

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/WorkflowV2Controller.java [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+     * @param loginUser             login user
+     * @param workflowCreateRequest the new workflow object will be created
+     * @return ResourceResponse object created
+     */
+    @ApiOperation(value = "create", notes = "CREATE_WORKFLOWS_NOTES")
+    @PostMapping(consumes = {"application/json"})
+    @ResponseStatus(HttpStatus.CREATED)
+    @ApiException(CREATE_PROCESS_DEFINITION_ERROR)
+    @AccessLogAnnotation(ignoreRequestArgs = "loginUser")
+    public ResourceResponse createWorkflows(@ApiIgnore @RequestAttribute(value = Constants.SESSION_USER) User loginUser,
```
Comment (ruanwenjun) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/WorkflowV2Controller.java]:
```suggestion
    public Result<ProcessDefinition> createWorkflows(@ApiIgnore @RequestAttribute(value = Constants.SESSION_USER) User loginUser,
```

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/WorkflowV2Controller.java [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+     * @param code             workflow resource code you want to update
+     * @param workflowUpdateRequest workflowUpdateRequest
+     * @return ResourceResponse object updated
+     */
+    @ApiOperation(value = "update", notes = "UPDATE_WORKFLOWS_NOTES")
+    @PutMapping(value = "/{code}")
+    @ResponseStatus(HttpStatus.OK)
+    @ApiException(UPDATE_PROCESS_DEFINITION_ERROR)
+    @AccessLogAnnotation(ignoreRequestArgs = "loginUser")
+    public ResourceResponse updateWorkflows(@ApiIgnore @RequestAttribute(value = Constants.SESSION_USER) User loginUser,
```
Comment (ruanwenjun) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/WorkflowV2Controller.java]:
```suggestion
    public Result<ProcessDefinition> updateWorkflows(@ApiIgnore @RequestAttribute(value = Constants.SESSION_USER) User loginUser,
```

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/WorkflowV2Controller.java [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+     * @param loginUser        login user
+     * @param code             workflow resource code you want to update
+     * @return ResourceResponse object get from condition
+     */
+    @ApiOperation(value = "get", notes = "GET_WORKFLOWS_NOTES")
+    @GetMapping(value = "/{code}")
+    @ResponseStatus(HttpStatus.OK)
+    @ApiException(QUERY_PROCESS_DEFINITION_LIST)
+    @AccessLogAnnotation(ignoreRequestArgs = "loginUser")
+    public ResourceResponse getWorkflows(@ApiIgnore @RequestAttribute(value = Constants.SESSION_USER) User loginUser,
```
Comment (ruanwenjun) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/WorkflowV2Controller.java]:
```suggestion
    public Result<ProcessDefinition> getWorkflows(@ApiIgnore @RequestAttribute(value = Constants.SESSION_USER) User loginUser,
```

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/WorkflowV2Controller.java [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+     * @param loginUser        login user
+     * @param workflowFilterRequest workflowFilterRequest
+     * @return PageResourceResponse from condition
+     */
+    @ApiOperation(value = "get", notes = "FILTER_WORKFLOWS_NOTES")
+    @GetMapping(consumes = {"application/json"})
+    @ResponseStatus(HttpStatus.OK)
+    @ApiException(QUERY_PROCESS_DEFINITION_LIST)
+    @AccessLogAnnotation(ignoreRequestArgs = "loginUser")
+    public PageResourceResponse filterWorkflows(@ApiIgnore @RequestAttribute(value = Constants.SESSION_USER) User loginUser,
```
Comment (ruanwenjun) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/WorkflowV2Controller.java]:
```suggestion
    public Result<PageInfo<ProcessDefinition>> filterWorkflows(@ApiIgnore @RequestAttribute(value = Constants.SESSION_USER) User loginUser,
```

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/dto/ResourceResponse.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+import org.apache.dolphinscheduler.api.enums.Status;
+import org.apache.dolphinscheduler.api.utils.Result;
+
+import lombok.Data;
+
+/**
+ * workflow response
+ */
+@Data
+public class ResourceResponse<T> extends Result<T> {
```
Comment (ruanwenjun) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/dto/ResourceResponse.java]: It seems this pojo is same with Result.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/dto/schedule/ScheduleCreateRequest.java (hunk 2 above)
Comment (ruanwenjun) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/dto/schedule/ScheduleCreateRequest.java:81]: It's better to create a pojo named ScheduleParam with this 4 fields.

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/SchedulerServiceImpl.java [hunk 10]
```diff
@@ -270,6 +350,61 @@ public Map<String, Object> updateSchedule(User loginUser,
...
+    /**
+     * get schedule object
+     *
+     * @param loginUser login user
+     * @param scheduleId scheduler id
+     * @return Schedule object
+     */
+    @Override
+    @Transactional
+    public Schedule getSchedules(User loginUser,
```
Comment (Tianqi-Dotes) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/SchedulerServiceImpl.java]: ->getSchedule

**Review by Tianqi-Dotes — DISMISSED** [review]

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/SchedulerServiceImpl.java (hunk 10 above)
Comment (zhongjiajie) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/SchedulerServiceImpl.java]: done

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/dto/ResourceResponse.java (hunk 9 above)
Comment (zhongjiajie) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/dto/ResourceResponse.java]: done

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/WorkflowV2Controller.java (hunk 8 above)
Comment (zhongjiajie) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/WorkflowV2Controller.java]: done

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/WorkflowV2Controller.java (hunk 7 above)
Comment (zhongjiajie) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/WorkflowV2Controller.java]: done

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java (hunk 4 above)
Comment (zhongjiajie) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java]: v1 is for currently API(operate workflow and task and task relation object), and V2 is only operate workflow object. V2 is mainly for RESTful api, which we only change one single object instand of too many object

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/dto/schedule/ScheduleUpdateRequest.java (hunk 3 above)
Comment (zhongjiajie) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/dto/schedule/ScheduleUpdateRequest.java]: done

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/dto/schedule/ScheduleCreateRequest.java (hunk 2 above)
Comment (zhongjiajie) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/dto/schedule/ScheduleCreateRequest.java:81]: > Why not use json util?
can you give some example about how to use it?

Comment (zhongjiajie) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/dto/schedule/ScheduleCreateRequest.java:81]: >It's better to create a pojo named ScheduleParam with this 4 fields.
@ruanwenjun  I want keep parameter same leve, as it in table `t_ds_scheldule`. If we add pojo, we have to accept one parameter like 

```json
{
    "releaseState": "Online",
    "ScheduleParam": "{'startTime': startTime, 'endTime': endTime, 'crontab': contab}"
}
```

but we I want it is accept parame like 

```json
{
    "releaseState": "Online",
    "startTime": "startTime",
    "endTime": "endTime",
    "crontab": "contab"
}
```

Comment (zhongjiajie) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/dto/schedule/ScheduleCreateRequest.java:81]: Oh I get your point, we can create pojo and then 

```java
ScheduleParam scheduleParam = new ScheduleParam(this.startTime, this.endTime, this.crontab, this.timezoneId);
return scheduleParam.toString();
```

right?

Comment (zhongjiajie) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/dto/schedule/ScheduleCreateRequest.java:81]: I currently use 

```java
Gson gson = new GsonBuilder().serializeNulls().create();
ScheduleParam scheduleParam = new ScheduleParam(this.startTime, this.endTime, this.crontab, this.timezoneId);
return gson.toJson(scheduleParam);
```

to covert pojo into json string

**Comment by zhongjiajie** [comment]
we still have discuss in https://github.com/apache/dolphinscheduler/pull/11912#discussion_r973823553 and https://github.com/apache/dolphinscheduler/pull/11912#discussion_r973825097

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java (hunk 4 above)
Comment (caishunfeng) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java]: Will we use v2 uniformly in the future? including UI or support restful api or pyds?

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/dto/schedule/ScheduleCreateRequest.java (hunk 2 above)
Comment (caishunfeng) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/dto/schedule/ScheduleCreateRequest.java:81]: > I currently use
> 
> ```java
> Gson gson = new GsonBuilder().serializeNulls().create();
> ScheduleParam scheduleParam = new ScheduleParam(this.startTime, this.endTime, this.crontab, this.timezoneId);
> return gson.toJson(scheduleParam);
> ```
> 
> to covert pojo into json string

I think it's OK, cc @ruanwenjun

**Review by caishunfeng — APPROVED** [review]
LGTM overall

**Comment by caishunfeng** [comment]
Thanks @zhongjiajie

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java (hunk 4 above)
Comment (zhongjiajie) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java]: I hope we can do that, and in that moment, maybe it it the best time to migrate python api into separate repository

**Pull request merged**